### PR TITLE
client: wait handleResourceTokenDispatcher exit

### LINKED ISSUE #6613: all, pd client: resource manager client got random 'null pointer dereference' panic on exit
## Enhancement Task
Since it's rare, I treat it as an enhancement task instead of a bug.

### what did I do?
Doing fault injection test with pd-tso-bench while pd clients were re-created with high frequency to test server side behavior.

---

2023/06/15 07:41:58.409 +00:00] [INFO] [pd_service_discovery.go:278] ["[pd] close pd service discovery client"]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x775288]

goroutine 560365 [running]:
github.com/tikv/pd/client.(*client).tryResourceManagerConnect(0x6?, {0xa78dd0, 0x4000b7ff40}, 0x4001731e70)
        /go/src/github.com/tikv/pd/client/resource_manager_client.go:354 +0x88
github.com/tikv/pd/client.(*client).handleResourceTokenDispatcher(0x4000bf3440, {0xa78dd0, 0x4000b7ff40}, 0x4000526c18)
        /go/src/github.com/tikv/pd/client/resource_manager_client.go:278 +0x44
created by github.com/tikv/pd/client.(*client).createTokenDispatcher
        /go/src/github.com/tikv/pd/client/resource_manager_client.go:265 +0x160

---

### root cause analysis
The pd client (resource manager client) exited without waiting for handleResourceTokenDispatcher() go routine to exit.

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: Close #6613

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has the configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

**Review by rleungx — APPROVED** [review]

File: client/resource_manager_client.go [hunk 1]
```diff
@@ -304,11 +304,13 @@ func (c *client) createTokenDispatcher() {
 		tokenBatchController: newTokenBatchController(
 			make(chan *tokenRequest, 1)),
 	}
+	c.wg.Add(1)
 	go c.handleResourceTokenDispatcher(dispatcherCtx, dispatcher.tokenBatchController)
 	c.tokenDispatcher = dispatcher
 }
 
 func (c *client) handleResourceTokenDispatcher(dispatcherCtx context.Context, tbc *tokenBatchController) {
+	defer c.wg.Done()
```
Comment (HuSharp) [inline client/resource_manager_client.go:313]: Is it better to add a log at `case <-dispatcherCtx.Done():` to indicate the exit

Comment (nolouch) [reply client/resource_manager_client.go:313]: cc @CabinfeverB

**Review by HuSharp — APPROVED** [review]
LGTM!

**Pull request merged**